README/Data.agda

```diff
@@ -0,0 +1,190 @@
+------------------------------------------------------------------------
+
+-- Basic datatypes are usually organised as follows:
+
+-- 1. A `Base` module which either contains the definition of the
+-- datatype or reimports it from the builtin modules, along with common
+-- functions, operations and relations over elements of the datatype.
+
+import Data.Nat.Base
+import Data.Integer.Base
+import Data.Char.Base
+import Data.String.Base
+import Data.Bool.Base
+
+-- Commonly these modules don't need to be imported directly as their
+-- contents is re-exported by the top level module (see below).
+
+-- 2. A `Properties` module which contains the basic properties of the
+-- functions, operations and relations contained in the base module.
+
+import Data.Nat.Properties
+import Data.Integer.Properties
+import Data.Char.Properties
+import Data.String.Properties
+import Data.Bool.Properties
+
+-- 3. A top-level module which re-exports the contents of the base
+-- module as well as various queries (i.e. decidability proofs) from the
+-- properties file.
+
+import Data.Nat
+import Data.Integer
+import Data.Char
+import Data.String
+import Data.Bool
+
+-- 4. A `Solver` module (for those datatypes that have an algebraic solver)
+-- which can be used to automatically solve equalities over the basic datatype.
+
+import Data.Nat.Solver
+import Data.Integer.Solver
+import Data.Bool.Solver
+
+-- 5. More complex operations and relations are commonly found in their
+-- own module beneath the top-level directory. For example:
+
+import Data.Nat.DivMod
+import Data.Integer.Coprimality
+
+-- Note that eventually there is a plan to re-organise the library to
+-- have the top-level module export a far wider range of properties and
+-- additional operations in order to minimise the number of imports
+-- needed. Currently it is necessary to import each of these seperately
+-- however.
+
+------------------------------------------------------------------------
+-- Container datatypes
+------------------------------------------------------------------------
+
+-- 1. As with basic datatypes, a `Base` module which contains the
+-- definition of the datatype, along with common functions and
+-- operations over that data. Unlike basic datatypes, the `Base` module
+-- for container datatypes does not export any relations or predicates
+-- over the datatype (see the `Relation` section below).
+
+import Data.List.Base
+import Data.Maybe.Base
+import Data.Sum.Base
+
+-- Commonly these modules don't need to be imported directly as their
+-- contents is re-exported by the top level module (see below).
+
+-- 2. As with basic datatypes, a `Properties` module which contains the
+-- basic properties of the functions, operations and relations contained
+-- in the base module.
+
+import Data.List.Properties
+import Data.Maybe.Properties
+import Data.Sum.Properties
+
+-- 3. As with basic datatypes, a top-level module which re-exports the
+-- contents of the base module.
+
```

Comment:
Should we mention that these can contain additional functions which could not possibly
fit into the corresponding `Base` module because of cyclic dependencies?